### PR TITLE
Development nextcloud weblet

### PR DESCRIPTION
**Pull request by Mik-TF**
Added the nextcloud all-in-one weblet on the playground based on this nextcloud flist: https://github.com/threefoldtech/tf-images/tree/development_nextcloudaio/tfgrid3/nextcloud

* The nextcloud flist works on the current main net playground.
* This weblet works when I build it locally on the Playground dev net

Note: I haven't tested it with RMB + internet as I couldn't make RMB work so far on my MAC.

There might be things to improve. Please let me know.

### CODE REVIEW

File: packages/playground/src/weblets/tf_nextcloud.vue [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+  <weblet-layout
+    ref="layout"
+    @mount="layoutMount"
+    :cpu="solution?.cpu"
+    :memory="solution?.memory"
+    :disk="disks.reduce((total, disk) => total + disk.size, rootFilesystemSize)"
+    :ipv4="ipv4"
+    :certified="certified"
+    :dedicated="dedicated"
+    title-image="images/icons/nextcloud.png"
```
Comment (MohamedElmdary) [inline packages/playground/src/weblets/tf_nextcloud.vue:11]:
```suggestion
  <weblet-layout
    ref="layout"
    :cpu="solution?.cpu"
    :memory="solution?.memory"
    :disk="disks.reduce((total, disk) => total + disk.size, rootFilesystemSize + (solution?.disk ?? 0))"
    :ipv4="ipv4"
    :certified="certified"
    :dedicated="dedicated"
    title-image="images/icons/nextcloud.png"
```

1. We don't need to use `layoutMount` here as will be shown later
2. We need to add disk size in count in order to update price calculator based on the new value `rootFilesystemSize + (solution?.disk ?? 0)`

File: packages/playground/src/weblets/tf_nextcloud.vue [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+  >
+    <template #title>Deploy Nextcloud All-in-One Instance </template>
+
+    <d-tabs :tabs="[{ title: 'Config', value: 'config' }]" ref="tabs">
+      <template #config>
+        <input-validator
```
Comment (MohamedElmdary) [inline packages/playground/src/weblets/tf_nextcloud.vue:17]:
```suggestion
<input-validator
```

As this weblet only 1 tab we don't need `d-tabs` component anymore

File: packages/playground/src/weblets/tf_nextcloud.vue [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+          >
+            <input-tooltip tooltip="Environment Value.">
+              <v-textarea label="Value" v-model="envs[index].value" no-resize :spellcheck="false" />
+            </input-tooltip>
+          </input-validator>
+        </ExpandableLayout>
+      </template>
+    </d-tabs>
+
+    <template #footer-actions>
```
Comment (MohamedElmdary) [inline packages/playground/src/weblets/tf_nextcloud.vue:123]:
```suggestion
        </SelectFarmManager>
      
    <template #footer-actions>
```

As I mentioned earlier we don't need tabs so we don't need to manage #envs tab anymore

File: packages/playground/src/weblets/tf_nextcloud.vue [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+            :filters="{
+              cpu: solution?.cpu,
+              memory: solution?.memory,
+              publicIp: ipv4,
+              ssd: disks.reduce((total, disk) => total + disk.size, rootFilesystemSize),
+              rentedBy: dedicated ? profileManager.profile?.twinId : undefined,
+              certified: certified,
+            }"
+            v-model="farm"
+          />
```
Comment (MohamedElmdary) [inline packages/playground/src/weblets/tf_nextcloud.vue:72]:
```suggestion
        <SelectFarmManager>
          <SelectFarm
            :filters="{
              cpu: solution?.cpu,
              memory: solution?.memory,
              publicIp: ipv4,
              ssd: rootFilesystemSize + (solution?.disk ?? 0),
              rentedBy: dedicated ? profileManager.profile?.twinId : undefined,
              certified: certified,
            }"
            v-model="farm"
          />
```

There is no disks to manage in this weblet

File: packages/playground/src/weblets/tf_nextcloud.vue [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+  envs.value.unshift({
+    key: "SSH_KEY",
+    value: profileManager.profile!.ssh,
+  });
+
+  envs.value.unshift({
+    key: "NEXTCLOUD_DOMAIN",
+    value: ipv4.value,
+  });
+}
```
Comment (MohamedElmdary) [inline packages/playground/src/weblets/tf_nextcloud.vue:180]:
```suggestion
const name = ref(generateName(8, { prefix: "nc" }));
const flist = ref<Flist>();
const ipv4 = ref(true);
const planetary = ref(true);
const wireguard = ref(true);
const farm = ref() as Ref<Farm>;
const network = ref();
const dedicated = ref(false);
const certified = ref(false);
const selectedNode = ref() as Ref<INode>;
const rootFilesystemSize = 8;
```

We don't need to manage envs or disks

File: packages/playground/src/weblets/tf_nextcloud.vue [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+          envs: envs.value,
+          planetary: planetary.value,
+          publicIpv4: ipv4.value,
+          rootFilesystemSize: rootFilesystemSize.value,
+          nodeId: selectedNode.value.nodeId,
+          rentedBy: dedicated.value ? grid!.twinId : undefined,
+          certified: certified.value,
+        },
+      ],
+    });
```
Comment (MohamedElmdary) [inline packages/playground/src/weblets/tf_nextcloud.vue:217]:
```suggestion
    const vm = await deployVM(grid!, {
      name: name.value,
      network: {
        addAccess: wireguard.value,
      },
      machines: [
        {
          name: name.value,
          cpu: solution.value.cpu,
          memory: solution.value.memory,
          flist: flist.value!.value,
          entryPoint: flist.value!.entryPoint,
          farmId: farm.value.farmID,
          farmName: farm.value.name,
          country: farm.value.country,
          disks: [{ size: solution.value.disk, mountPoint: "/mnt/next_cloud" }], // make sure to update mountPoint as needed
          envs: [
            { key: "SSH_KEY", value: profileManager.profile!.ssh },
            { key: "NEXTCLOUD_DOMAIN", value: String(ipv4.value) },
          ],
          planetary: planetary.value,
          publicIpv4: ipv4.value,
          rootFilesystemSize: rootFilesystemSize, // remove `.value`
          nodeId: selectedNode.value.nodeId,
          rentedBy: dedicated.value ? grid!.twinId : undefined,
          certified: certified.value,
        },
      ],
    });
```

**Comment by AhmedHanafy725** [comment]
the changes on api docs of the grid client are not related to this PR, also it's hard to review. It's auto-generated.

so can you remove them from this PR? and I will regenerate them, then you can check if the changes you need are done or not

**Comment by Mik-TF** [comment]
@MohamedElmdary 
Thank you so much Mohamed!

It's my first time implementing a weblet so I was mostly in experimental mode. It helped a great lot to say the least to have your suggestions.

@AhmedHanafy725 Thank you too Ahmed! What I did is that I replaced the 5 .html files with the latest dev version you just uploaded.

It never was my intent to change those documents. I tried to do some more precise git add <file-name> after reading you here.

***

There are still some errors in the Full Clients Lint test so I will try to check what to fix.

Feel free to check it too since you guys are in the known compared to me!

Thanks again.

File: packages/playground/src/weblets/tf_nextcloud.vue (hunk 4 above)
Comment (Mik-TF) [reply packages/playground/src/weblets/tf_nextcloud.vue]: Just to be sure I understand correctly: when I deploy [the Nextcloud FList](https://github.com/threefoldtech/tf-images/tree/development_nextcloudaio/tfgrid3/nextcloud) on the playground, I set 8gb of root file system for the nextcloud's packages, and I add a disk for the remaining space for Nextcloud storage.

In this case, aren't there the storage filesystem + the storage disk implied?

**Comment by Mik-TF** [comment]
@MohamedElmdary @AhmedHanafy725 

The only error left is at line 69 of tf_nextcloud.vue and it concerns the type for rootfilesystem.

**Comment by Mik-TF** [comment]
@MohamedElmdary 

If you could please have a look at the tf_nextcloud.vue file. I get this error:

> Error: src/weblets/tf_nextcloud.vue(76,15): error TS2322: Type '{ farmId: number; cpu: number; memory: number; ipv4: boolean; disks: { size: number; mountPoint: string; }[]; rentedBy: number | undefined; certified: boolean; }' is not assignable to type 'NodeFilters'.
  Object literal may only specify known properties, and 'disks' does not exist in type 'NodeFilters'.
  
And I can't find the fix after different tries.

Thanks!
Let me know if I can help. I ran out of ideas for now.

**Comment by Mik-TF** [comment]
I've just seen the changes in the [last PR](https://github.com/threefoldtech/tfgrid-sdk-ts/pull/937/files)

I'll give this a try.

**Comment by Mik-TF** [comment]
Continuing this work in a new PR #960

**Pull request closed without merging**